src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -422,6 +424,17 @@ private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F,
                     !property.IsStatic &&
                     (property.ParameterCount == 0))
                 {
+                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                    Conversion conversion = F.Compilation.Conversions.ClassifyConversionFromType(property.Type, returnType, ref useSiteDiagnostics);
+                    if (conversion.Kind == ConversionKind.NoConversion)
+                    {
+                        var badTaskProperty = new CSDiagnostic(
+                            new CSDiagnosticInfo(ErrorCode.ERR_BadAsyncMethodBuilderTaskProperty, builderType, returnType, property.Type),
+                            F.Syntax.Location);
+                        F.Diagnostics.Add(badTaskProperty);
+                        return null;
```

Comment:
I noticed that. Adding a specific error message is cheap and low-risk. I'd probably keep it, if that's ok.

By the way, I wonder if the text for `error CS1983: The return type of an async method must be void, Task or Task<T>` should be updated to mention task-like types. As it stands, it seems pretty misleading. Maybe file an issue?